src/panels/BasePanel.ts

```diff
@@ -0,0 +1,129 @@
+ */
+export abstract class BasePanel<TInitialState, TToWebviewCommands, TToVsCodeCommands> {
+    protected constructor(
+        readonly extensionUri: Uri,
+        readonly contentId: string
+    ) { }
+
+    show(dataProvider: PanelDataProvider<TInitialState, TToWebviewCommands, TToVsCodeCommands>) {
+        const panelOptions = {
+            enableScripts: true,
+            // Restrict the webview to only load resources from the `webview-ui/dist` directory
+            localResourceRoots: [Uri.joinPath(this.extensionUri, "webview-ui/dist")],
+        };
+
+        const title = dataProvider.getTitle();
+
+        const panel = window.createWebviewPanel(viewType, title, ViewColumn.One, panelOptions);
+        const disposables: Disposable[] = [];
+
+        // Set up messaging between VSCode and the webview.
+        const messageContext = new WebviewMessageContext<TToWebviewCommands, TToVsCodeCommands>(panel.webview, disposables);
+        const subscriber = dataProvider.createSubscriber(messageContext);
+        if (subscriber) {
+            messageContext.subscribeToMessages(subscriber);
+        }
+
+        // Set an event listener to listen for when the panel is disposed (i.e. when the user closes
+        // the panel or when the panel is closed programmatically)
+        panel.onDidDispose(() => this._dispose(panel, disposables), null, disposables);
+
+        // Set the HTML content for the webview panel
+        const initialState = dataProvider.getInitialState();
+        panel.webview.html = this._getWebviewContent(panel.webview, this.extensionUri, title, initialState);
+    }
+
+    private _dispose(panel: WebviewPanel, disposables: Disposable[]) {
+        panel.dispose();
+        disposables.forEach(d => d.dispose());
+    }
+
+    private _getWebviewContent(webview: Webview, extensionUri: Uri, title: string, initialState: TInitialState | undefined) {
+        // Get URIs for the React build output.
+        const stylesUri = getUri(webview, extensionUri, ["assets", "index.css"]);
+        const scriptUri = getUri(webview, extensionUri, ["assets", "index.js"]);
+
+        // See: https://github.com/microsoft/vscode-webview-ui-toolkit/blob/main/docs/getting-started.md#enable-webview-scripts-and-improve-security
+        const nonce = getNonce();
+
+        const initialStateJson = initialState ? JSON.stringify(initialState) : "";
+
+        // Tip: Install the es6-string-html VS Code extension to enable code highlighting below
+        return /*html*/ `
+        <!DOCTYPE html>
+        <html lang="en">
+            <head>
+                <meta charset="UTF-8" />
+                <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+                <meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src ${webview.cspSource} https://use.fontawesome.com; script-src 'nonce-${nonce}' https://use.fontawesome.com; font-src 'self' https://use.fontawesome.com">
```

Comment:
I think we can/should consume the [fontawesome npm package](https://www.npmjs.com/package/font-awesome) in the `webview-ui` project, rather than accessing an external URL here. But perhaps save that for another day.